apps/src/code-studio/initApp/project.js

```diff
@@ -699,6 +700,29 @@ var projects = module.exports = {
   clearHtml() {
     currentSources.html = '';
   },
+  /**
+   * Saves the project only if the sources {source, html, animations,
+   * makerAPIsEnabled} have changed.
+   * @returns {Promise} A promise containing the project data if the project
+   * was saved, otherwise returns a promise which resolves with no arguments.
+   */
+  saveIfSourcesChanged() {
+    if (!isEditable()) {
+      return Promise.resolve();
+    }
+
+    return new Promise(resolve => {
+      this.getUpdatedSourceAndHtml_(newSources => {
+        const sourcesChanged = (JSON.stringify(currentSources) !== JSON.stringify(newSources));
+        if (sourcesChanged || thumbnailChanged) {
+          thumbnailChanged = false;
+          this.saveSourceAndHtml_(newSources, resolve);
+        } else {
+          resolve();
+        }
+      });
+    });
+  },
```

Comment:
:grin: This makes me want to suggest `async` / `await` since Chris just added support... but let's not introduce any more complexity into the project system _just_ yet.